File: jetpack/lib-arch/src/main/java/com/yyxnb/what/arch/lifecycle/FragmentLifecycle.java
package com.yyxnb.what.arch.lifecycle;

import android.content.Context;
import android.os.Bundle;
import android.util.LruCache;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import com.yyxnb.what.arch.helper.ActivityHelper;

/**
 * ================================================
 * 作    者：yyx
 * 版    本：1.0
 * 日    期：2020/11/21
 * 历    史：
 * 描    述：Fragment 注册生命周期监听
 * ================================================
 */
public class FragmentLifecycle extends FragmentManager.FragmentLifecycleCallbacks {

    private static volatile FragmentLifecycle mInstance = null;

    private FragmentLifecycle() {
    }

    public static FragmentLifecycle getInstance() {
        if (null == mInstance) {
            synchronized (FragmentLifecycle.class) {
                if (null == mInstance) {
                    mInstance = new FragmentLifecycle();
                }
            }
        }
        return mInstance;
    }

    private final LruCache<String, IFragmentLifecycle> cache = new LruCache<>(100);

    @Override
    public void onFragmentAttached(@NonNull FragmentManager fm, @NonNull Fragment f, @NonNull Context context) {
        ActivityHelper.getInstance().addFragment(f);
        IFragmentLifecycle fragmentDelegate = fetchFragmentDelegate(f, fm);
        fragmentDelegate.onAttached(context);
    }

    @Override
    public void onFragmentCreated(@NonNull FragmentManager fm, @NonNull Fragment f, @Nullable Bundle savedInstanceState) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onCreated(savedInstanceState);
        }
    }

    @Override
    public void onFragmentViewCreated(@NonNull FragmentManager fm, @NonNull Fragment f, @NonNull View v, @Nullable Bundle savedInstanceState) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onViewCreated(v, savedInstanceState);
        }
    }

    @Override
    public void onFragmentActivityCreated(@NonNull FragmentManager fm, @NonNull Fragment f, @Nullable Bundle savedInstanceState) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onActivityCreated(savedInstanceState);
        }
    }

    @Override
    public void onFragmentStarted(@NonNull FragmentManager fm, @NonNull Fragment f) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onStarted();
        }
    }

    @Override
    public void onFragmentResumed(@NonNull FragmentManager fm, @NonNull Fragment f) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onResumed();
        }
    }

    @Override
    public void onFragmentPaused(@NonNull FragmentManager fm, @NonNull Fragment f) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onPaused();
        }
    }

    @Override
    public void onFragmentStopped(@NonNull FragmentManager fm, @NonNull Fragment f) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onStopped();
        }
    }

    @Override
    public void onFragmentSaveInstanceState(@NonNull FragmentManager fm, @NonNull Fragment f, @NonNull Bundle outState) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onSaveInstanceState(outState);
        }
    }

    @Override
    public void onFragmentViewDestroyed(@NonNull FragmentManager fm, @NonNull Fragment f) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onViewDestroyed();
        }
    }

    @Override
    public void onFragmentDestroyed(@NonNull FragmentManager fm, @NonNull Fragment f) {
        ActivityHelper.getInstance().removeFragment(f);
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onDestroyed();
        }
    }

    @Override
    public void onFragmentDetached(@NonNull FragmentManager fm, @NonNull Fragment f) {
        if (fetchFragmentDelegateFromCache(f) != null) {
            fetchFragmentDelegateFromCache(f).onDetached();
        }
        cache.remove(getKey(f));
    }

    private IFragmentLifecycle fetchFragmentDelegate(Fragment fragment, FragmentManager manager) {

        IFragmentLifecycle fragmentDelegate;

        if (fetchFragmentDelegateFromCache(fragment) != null) {
            fragmentDelegate = fetchFragmentDelegateFromCache(fragment);
        } else {
            fragmentDelegate = newDelegate(manager, fragment);
        }
        cache.put(getKey(fragment), fragmentDelegate);
        return fragmentDelegate;
    }

    private IFragmentLifecycle fetchFragmentDelegateFromCache(Fragment fragment) {
        return cache.get(getKey(fragment));
    }

    private IFragmentLifecycle newDelegate(FragmentManager manager, Fragment fragment) {
        return new FragmentLifecycleImpl(fragment, manager);
    }

    private String getKey(Fragment fragment) {
        return fragment.getClass().getName() + fragment.hashCode();
    }

}

File: arch/src/main/java/com/yyxnb/arch/delegate/FragmentLifecycle.kt
package com.yyxnb.arch.delegate

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.view.View
import com.yyxnb.arch.interfaces.IFragment
import com.yyxnb.arch.interfaces.IFragmentLifecycle

object FragmentLifecycle : FragmentManager.FragmentLifecycleCallbacks() {

    private val cacheDelegate by lazy { HashMap<String, IFragmentLifecycle>() }

    private lateinit var IFragmentLifecycle: IFragmentLifecycle

    override fun onFragmentAttached(fm: FragmentManager, f: Fragment, context: Context) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onAttached(context) }
    }

    override fun onFragmentCreated(fm: FragmentManager, f: Fragment, savedInstanceState: Bundle?) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onCreated(savedInstanceState) }
    }

    override fun onFragmentViewCreated(fm: FragmentManager, f: Fragment, v: View, savedInstanceState: Bundle?) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onViewCreated(v, savedInstanceState) }
    }

    override fun onFragmentActivityCreated(fm: FragmentManager, f: Fragment, savedInstanceState: Bundle?) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onActivityCreate(savedInstanceState) }
    }

    override fun onFragmentStarted(fm: FragmentManager, f: Fragment) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onStarted() }
    }

    override fun onFragmentResumed(fm: FragmentManager, f: Fragment) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onResumed() }
    }

    override fun onFragmentPaused(fm: FragmentManager, f: Fragment) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onPaused() }
    }

    override fun onFragmentStopped(fm: FragmentManager, f: Fragment) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onStopped() }
    }

    override fun onFragmentSaveInstanceState(fm: FragmentManager, f: Fragment, outState: Bundle) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onSaveInstanceState(outState) }
    }

    override fun onFragmentViewDestroyed(fm: FragmentManager, f: Fragment) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onViewDestroyed() }
    }

    override fun onFragmentDestroyed(fm: FragmentManager, f: Fragment) {
        forwardDelegateFunction(fm, f) { IFragmentLifecycle.onDestroyed() }
    }

    override fun onFragmentDetached(fm: FragmentManager, f: Fragment) {
        forwardDelegateFunction(fm, f) {
            IFragmentLifecycle.onDetached()
            cacheDelegate.clear()
        }
    }

    private fun forwardDelegateFunction(fm: FragmentManager, f: Fragment, block: () -> Unit) {

        if (f !is IFragment) return

        if (!this::IFragmentLifecycle.isInitialized || !IFragmentLifecycle.isAdded()) {

            val key = f.javaClass.name

            IFragmentLifecycle = cacheDelegate[key] ?: newDelegate(fm, f, key)
        }

        block()
    }

    private fun newDelegate(fm: FragmentManager, f: Fragment, key: String): IFragmentLifecycle {
        return realNewDelegate(fm, f).also { cacheDelegate[key] = it }
    }

    private fun realNewDelegate(fm: FragmentManager, f: Fragment): IFragmentLifecycle {
        return FragmentLifecycleDelegate(fm, f)
    }

}